### PR TITLE
Fix skiptoken paging issue when combined with $orderby involving nullable property

### LINKED ISSUE #1943: $Skiptoken pagination issue when combined with $orderby
Pagination is not working when it is based on the $skiptoken and combined with $orderby on properties of type Byte or nullable integer (int?).

I get this error on OData WebApi lib 7.2.1 and 7.2.2

You can reproduce the issue using the solution available [here](https://github.com/MYoucef/OData-skiptoken-issue/tree/master/skiptoken)

The whole error message is:
```json
{
    "error": {
        "code": "",
        "message": "The query specified in the URI is not valid. The binary operator AndAlso is not defined for the types 'System.Nullable`1[System.Boolean]' and 'System.Boolean'.",
        "innererror": {
            "message": "The binary operator AndAlso is not defined for the types 'System.Nullable`1[System.Boolean]' and 'System.Boolean'.",
            "type": "System.InvalidOperationException",
            "stacktrace": "   at System.Linq.Expressions.Expression.AndAlso(Expression left, Expression right, MethodInfo method)\r\n   at Microsoft.AspNet.OData.Query.DefaultSkipTokenHandler.ApplyToCore(IQueryable query, ODataQuerySettings querySettings, IList`1 orderByNodes, ODataQueryContext context, String skipTokenRawValue)\r\n   at Microsoft.AspNet.OData.Query.DefaultSkipTokenHandler.ApplyTo(IQueryable query, SkipTokenQueryOption skipTokenQueryOption)\r\n   at Microsoft.AspNet.OData.Query.SkipTokenQueryOption.ApplyTo(IQueryable query, ODataQuerySettings querySettings, ODataQueryOptions queryOptions)\r\n   at Microsoft.AspNet.OData.Query.ODataQueryOptions.ApplyTo(IQueryable query, ODataQuerySettings querySettings)\r\n   at Microsoft.AspNet.OData.EnableQueryAttribute.ApplyQuery(IQueryable queryable, ODataQueryOptions queryOptions)\r\n   at Microsoft.AspNet.OData.EnableQueryAttribute.ExecuteQuery(Object responseValue, IQueryable singleResultCollection, IWebApiActionDescriptor actionDescriptor, Func`2 modelFunction, IWebApiRequestMessage request, Func`2 createQueryOptionFunction)\r\n   at Microsoft.AspNet.OData.EnableQueryAttribute.OnActionExecuted(Object responseValue, IQueryable singleResultCollection, IWebApiActionDescriptor actionDescriptor, IWebApiRequestMessage request, Func`2 modelFunction, Func`2 createQueryOptionFunction, Action`1 createResponseAction, Action`3 createErrorAction)"
        }
    }
}

### LINKED ISSUE #2041: Error when SkipToken contains a nullable string
If sorting by a nullable string, `$skiptoken` may contain a null value. This causes an exception show below.

Example URL:

```
https://localhost:44322/odata/TransactionExceptionDetails?$select=TxnId%2CContactCity%2CContactState&$orderby=ContactCity%2CTxnId&$skiptoken=ContactCity-null,TxnId-12005
```

Note the use of `$orderby=ContactCity%2CTxnId` and `$skiptoken=ContactCity-null,TxnId-12005`.

If I remove `ContactCity` from the order by, then the generated skiptoken doesn't contain it either and I don't get an error.

### Assemblies affected

```
    <PackageReference Include="Microsoft.AspNetCore.OData" Version="7.3.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.1" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.1" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="3.1.1">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="3.1.0" />
    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="3.1.0" />
```

### Reproduce steps

*The simplest set of steps to reproduce the issue. If possible, reference a commit that demonstrates the issue.*

### Expected result

*What would happen if there wasn't a bug.*

### Actual result

```
ArgumentException: Expression of type 'Microsoft.OData.ODataNullValue' cannot be used for parameter of type 'System.String' of method 'Int32 Compare(System.String, System.String, System.StringComparison)' (Parameter 'arg1')


System.Dynamic.Utils.ExpressionUtils.ValidateOneArgument(MethodBase method, ExpressionType nodeKind, Expression arguments, ParameterInfo pi, string methodParamName, string argumentParamName, int index)
System.Linq.Expressions.Expression.Call(MethodInfo method, Expression arg0, Expression arg1, Expression arg2)
Microsoft.AspNet.OData.Query.Expressions.ExpressionBinderBase.CreateBinaryExpression(BinaryOperatorKind binaryOperator, Expression left, Expression right, bool liftToNull)
Microsoft.AspNet.OData.Query.DefaultSkipTokenHandler.ApplyToCore(IQueryable query, ODataQuerySettings querySettings, IList<OrderByNode> orderByNodes, ODataQueryContext context, string skipTokenRawValue)
Microsoft.AspNet.OData.Query.DefaultSkipTokenHandler.ApplyTo(IQueryable query, SkipTokenQueryOption skipTokenQueryOption)
Microsoft.AspNet.OData.Query.SkipTokenQueryOption.ApplyTo(IQueryable query, ODataQuerySettings querySettings, ODataQueryOptions queryOptions)
Microsoft.AspNet.OData.Query.ODataQueryOptions.ApplyTo(IQueryable query, ODataQuerySettings querySettings)
Microsoft.AspNet.OData.EnableQueryAttribute.ApplyQuery(IQueryable queryable, ODataQueryOptions queryOptions)
Microsoft.AspNet.OData.EnableQueryAttribute.ExecuteQuery(object responseValue, IQueryable singleResultCollection, IWebApiActionDescriptor actionDescriptor, Func<Type, IEdmModel> modelFunction, IWebApiRequestMessage request, Func<ODataQueryContext, ODataQueryOptions> createQueryOptionFunction)
Microsoft.AspNet.OData.EnableQueryAttribute.OnActionExecuted(object responseValue, IQueryable singleResultCollection, IWebApiActionDescriptor actionDescriptor, IWebApiRequestMessage request, Func<Type, IEdmModel> modelFunction, Func<ODataQueryContext, ODataQueryOptions> createQueryOptionFunction, Action<HttpStatusCode> createResponseAction, Action<HttpStatusCode, string, Exception> createErrorAction)
Microsoft.AspNet.OData.EnableQueryAttribute.OnActionExecuted(ActionExecutedContext actionExecutedContext)
Microsoft.AspNetCore.Mvc.Filters.ActionFilterAttribute.OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

### LINKED ISSUE #2561: $skiptoken is always returns null when combined with $Select
*We are using Microsoft.AspNetCore.OData 7.5.6.
We have a API which returns a list of items with pagesize as 20, and skiptoken enabled.
The API returns response with nextlink with proper skiptoken when no $Select parameter is there.
whenever $Select parameter is added, it always return null for skiptoken *

### Assemblies affected

*Microsoft.AspNetCore.OData 7.5.6

### Reproduce steps

*Expose a API endpoint in ASP.Net Core with OData Support. Set page size, enable skiptoken as well.
When directly accessing the API endpoint without any oData Query then the next page link coming properly with correct skiptoken=id-1
But when we use $Select odata query , then the $skiptoken=id-null is returned.*

### Expected result

*when we use $Select odata query , then the $skiptoken should come proper*

### Actual result

*when we use $Select odata query , then the $skiptoken=id-null is returned.*

### PR DESCRIPTION
**Pull request by gathogojr**
### Issues

*This pull request fixes #1943, fixes #2041, and fixes #2561*

## Description
Consider a simple scenario involving a data sample comprising of `Id` key field and a non-unique nullable field `CreditLimit`:
| Id | CreditLimit |
| -- | ------------ |
| 1 | null |
| 2 | 2 |
| 3 | null |
| 4 | 30 |
| 5 | null |
| 6 | 35 |
| 7 | 5 |
| 8 | 50 |
| 9 | 25 |

From the controller action, we could configure a page size of 2:
```csharp
[EnableQuery(PageSize = 2)]
public ActionResult<IEnumerable<S1Customer>> Get()
{
    return new List<S1Customer>
    {
        new S1Customer { Id = 1, CreditLimit = null },
        new S1Customer { Id = 2, CreditLimit = 2 },
        new S1Customer { Id = 3, CreditLimit = null },
        new S1Customer { Id = 4, CreditLimit = 30 },
        new S1Customer { Id = 5, CreditLimit = null },
        new S1Customer { Id = 6, CreditLimit = 35 },
        new S1Customer { Id = 7, CreditLimit = 5 },
        new S1Customer { Id = 8, CreditLimit = 50 },
        new S1Customer { Id = 9, CreditLimit = 25 },
    };
}
```

When the data sample is ordered by `CreditLimit` field (`$orderby=CreditLimit`), and then internally by `Id` (for uniqueness and predictability - stable ordering), the ordered data sample would look at follows:
| CreditLimit | Id |
| ------------ | -- |
| null | 1 |
| null | 3 |
| null | 5 |
| 2 | 2 |
| 5 | 7 |
| 25 | 9 |
| 30 | 4 |
| 35 | 6 |
| 50 | 8 |

If we query for the `S1Customers` entity set and order by `CreditLimit`, we get the following response:
```http
GET: http://localhost:5000/odata/S1Customers?$orderby=CreditLimit
```
#### Response:
```json
{
    "@odata.context": "http://localhost:5000/odata/$metadata#S1Customers",
    "value": [
        {
            "Id": 1,
            "CreditLimit": null
        },
        {
            "Id": 3,
            "CreditLimit": null
        }
    ],
    "@odata.nextLink": "http://localhost:5000/odata/S1Customers?$orderby=CreditLimit&$skiptoken=CreditLimit-null,Id-3"
}
```

While the response is correct (including the next-link), if we attempt to use the returned next link to fetch the next batch of records, the following error message is returned:
```
The query specified in the URI is not valid. The binary operator GreaterThan is not defined for the types 'System.Nullable`1[System.Decimal]' and 'Microsoft.OData.ODataNullValue'.
```
What is happening here is that we’re mishandling the null value and attempting to apply a greater-than operator between the nullable `CreditLimit` field and the `ODataNullValue`.

This pull request fixes this issue by ensuring that a valid `Where` expression is composed as follows:
- for skiptoken value **`CreditLimit-null,Id-3`** in ascending order scenario - `$orderby=CreditLimit`, we compose the `Where` expression as: **`(CreditLimit ne null OR (CreditLimit eq null AND Id gt 3))`**
- for skiptoken value **`CreditLimit-2,Id-2`** in ascending order scenario - `$orderby=CreditLimit`, we compose the `Where` expression as: **`(CreditLimit gt 2 OR (CreditLimit eq 2 AND Id gt 2))`**
- for skiptoken value **`CreditLimit-null,Id-3`** in descending order scenario - `$orderby=CreditLimit desc`, we compose the `Where` expression as: **`(CreditLimit eq null AND Id gt 3)`**
- for skiptoken value **`CreditLimit-35,Id-6`** in descending order scenario - `$orderby=CreditLimit desc`, we compose the `Where` expression as: **`(CreditLimit lt 35 OR (CreditLimit eq 35 AND Id gt 6))`**

Consider also a more advanced scenario involving a data sample comprising of `Id` key field, a non-nullable field `Grade`, and a non-unique nullable field `CreditLimit`:
| Id | Grade | CreditLimit |
| -- | ----- | ------------ |
| 1 | A | null |
| 2 | B | null |
| 3 | A | 10 |
| 4 | C | null |
| 5 | A | 30 |
| 6 | C | null |
| 7 | B | 5 |
| 8 | C | 25 |
| 9 | B | 50 |
| 10 | D | 50 |
| 11 | F | 35 |
| 12 | F | 30 |
| 13 | F | 55 |

From the controller action, we would configure a page size of 4 as follows:
```csharp
[EnableQuery(PageSize = 4)]
public ActionResult<IEnumerable<S2Customer>> Get()
{
    return new List<S2Customer>
    {
        new S2Customer { Id = 1, Grade = "A", CreditLimit = null },
        new S2Customer { Id = 2, Grade = "B", CreditLimit = null },
        new S2Customer { Id = 3, Grade = "A", CreditLimit = 10 },
        new S2Customer { Id = 4, Grade = "C", CreditLimit = null },
        new S2Customer { Id = 5, Grade = "A", CreditLimit = 30 },
        new S2Customer { Id = 6, Grade = "C", CreditLimit = null },
        new S2Customer { Id = 7, Grade = "B", CreditLimit = 5 },
        new S2Customer { Id = 8, Grade = "C", CreditLimit = 25 },
        new S2Customer { Id = 9, Grade = "B", CreditLimit = 50 },
        new S2Customer { Id = 10, Grade = "D", CreditLimit = 50 },
        new S2Customer { Id = 11, Grade = "F", CreditLimit = 35 },
        new S2Customer { Id = 12, Grade = "F", CreditLimit = 30 }
    };
}
```

When the data sample is ordered by `Grade` and then by `CreditLimit` field ($orderby=Grade,CreditLimit), and then internally by `Id` (for uniqueness and predictability - stable ordering), the ordered data sample would look at follows:
| Grade | CreditLimit | Id |
| ----- | ----------- | -- |
| A | null | 1 |
| A | 10 | 3 |
| A | 30 | 5 |
| B | null | 2 |
| B | 5 | 7 |
| B | 50 | 9 |
| C | null | 4 |
| C | null | 6 |
| C | 25 | 8 |
| D | 50 | 10 |
| F | 30 | 12 |
| F | 35 | 11 |
| F | 55 | 13 |

If we query for the `S2Customers` entity set and orderby `Grade` and then by `CreditLimit`, we get the following response:
```http
GET: http://localhost:5000/odata/S2Customers?$orderby=Grade,CreditLimit
```
#### Response
```json
{
    "@odata.context": "http://localhost:5000/odata/$metadata#S2Customers",
    "value": [
        {
            "Id": 1,
            "Grade": "A",
            "CreditLimit": null
        },
        {
            "Id": 3,
            "Grade": "A",
            "CreditLimit": 10.00
        },
        {
            "Id": 5,
            "Grade": "A",
            "CreditLimit": 30.00
        },
        {
            "Id": 2,
            "Grade": "B",
            "CreditLimit": null
        }
    ],
    "@odata.nextLink": "http://localhost:5000/odata/S2Customers?$orderby=Grade%2CCreditLimit&$skiptoken=Grade-%27B%27,CreditLimit-null,Id-2"
}
```

Again, the response is correct (including the next-link), but if we attempt to use the returned next link to fetch the next batch of records, the following error message is returned:
```
The query specified in the URI is not valid. The binary operator GreaterThan is not defined for the types 'System.Nullable`1[System.Decimal]' and 'Microsoft.OData.ODataNullValue'.
```

Again, this is due to mishandling of the null value.

In this pull request we ensure that a valid `Where` expression is composed as follows:
- for skiptoken value of **`Grade-B,CreditLimit-null,Id-2`** in ascending order scenario - `$orderby=Grade,CreditLimit`, we compose the `Where` expression as: **`Grade gt B OR (Grade eq B AND (CreditLimit ne null OR (CreditLimit eq null AND Id gt 2)))`**
- for skiptoken value of **`Grade-F,CreditLimit-30,Id-12`** in ascending order scenario - `$orderby=Grade,CreditLimit`, we compose the `Where` expression as: **`Grade gt F OR (Grade eq F AND (CreditLimit gt 30 OR (CreditLimit eq 30 AND Id gt 12)))`**
- for skiptoken value of **`CreditLimit-null,Grade-C,Id-6`** in ascending order scenario - `$orderby=CreditLimit,Grade`, we compose the `Where` expression as: **`CreditLimit ne null OR (CreditLimit eq null AND (Grade gt C OR (Grade eq C AND Id gt 6)))`**
- for skiptoken value of **`CreditLimit-30,Grade-A,Id-5`** in ascending order scenario - `$orderby=CreditLimit,Grade`, we compose the `Where` expression as: **`CreditLimit gt 30 OR (CreditLimit eq 30 AND (Grade gt A OR (Grade eq A AND Id gt 5)))`**

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Comment by shdim** [comment]
It is still broken if $orderby nullable DateTime property and parameterized constants.

` Expression condition = binder.CreateBinaryExpression(
                            binaryOperator: BinaryOperatorKind.Equal,
                            left: property,
                            right: parameterizeConstant ? LinqParameterContainer.Parameterize(propertyType, null) : Expression.Constant(null),
                            liftToNull: false);`

LinqParameterContainer.Parameterize(propertyType, null) is a member expression with value null but ExpressionBinderBase.DateTimeOffsetToDateTime expects it to be either not null or a ConstantExpression.

`
            if(dto != null)
            {
                expressionValue = dto.Value;
            }
            else
            {
                expressionValue = (expression as ConstantExpression).Value;
            }
`
So there is a NullReferenceException at ` (expression as ConstantExpression).Value;`

**Comment by gathogojr** [comment]
> It is still broken if $orderby nullable DateTime property and parameterized constants.
> 
> ` Expression condition = binder.CreateBinaryExpression( binaryOperator: BinaryOperatorKind.Equal, left: property, right: parameterizeConstant ? LinqParameterContainer.Parameterize(propertyType, null) : Expression.Constant(null), liftToNull: false);`
> 
> LinqParameterContainer.Parameterize(propertyType, null) is a member expression with value null but ExpressionBinderBase.DateTimeOffsetToDateTime expects it to be either not null or a ConstantExpression.
> 
> `if(dto != null) { expressionValue = dto.Value; } else { expressionValue = (expression as ConstantExpression).Value; }` So there is a NullReferenceException at ` (expression as ConstantExpression).Value;`

Hi @shdim. Thank you for your feedback. I'll take a look and revert

**Comment by shdim** [comment]
> I have tweaked the logic a bit to handle that edge case. I have also added a couple of tests for `DateTime` scenario. Please give it another try when you get a chance.

Thank you @gathogojr, it runs now without problems for all my scenarios.

File: src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs [hunk 1]
```diff
@@ -252,40 +252,107 @@ private static IQueryable ApplyToCore(IQueryable query, ODataQuerySettings query
...
+
+                        // When value is NOT null in the ascending order scenario,
+                        // we are aiming for the following expression:
+                        // (Prop1 > Value1) OR ((Prop1 == Value1) AND (Prop2 > Value2)) ...
+
+                        compare = binder.CreateBinaryExpression(
+                            binaryOperator: BinaryOperatorKind.GreaterThan,
+                            left: property,
+                            right: constant,
+                            liftToNull: propertyIsNullable ? false : true);
```
Comment (habbes) [inline src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs:332]:
```suggestion
                            liftToNull: !propertyIsNullable);
```

File: src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs [hunk 2]
```diff
@@ -231,7 +231,7 @@ private static IQueryable ApplyToCore(IQueryable query, ODataQuerySettings query
                 directionMap = new Dictionary<string, OrderByDirection>();
             }
 
-            IDictionary<string, object> propertyValuePairs = PopulatePropertyValuePairs(skipTokenRawValue, context);
+            IDictionary<string, Tuple<object, Type>> propertyValuePairs = PopulatePropertyValuePairs(skipTokenRawValue, context);
```
[the remark this answers is not in the document]
Comment (gathogojr) [reply src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs:236]: @habbes The ASP.NET Classic project targets .NET45 that doesn't support `ValueTuple`

**Review by xuzhg — APPROVED** [review]

**Review by ElizabethOkerio — APPROVED** [review]

File: src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs:236]: Just seen the ValueTuple can be suppported in net4.5 by referencing the `System.ValueTuple NuGet Package`

[the remark this answers is not in the document]
Comment (gathogojr) [reply src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs:236]: @ElizabethOkerio True. I was aware, but I didn't want to introduce an extra dependency for a framework that I'd expect to be sparingly used currently
